src/frontends/lean/widget.h

```diff
@@ -0,0 +1,138 @@
+    friend vdom;
+public:
+  vdom_cell() : m_rc(0) {}
+  virtual ~vdom_cell() {}
+  virtual json to_json(list<unsigned> const &) { return json(); };
+  virtual optional<std::string> key() { return optional<std::string>(); };
+  virtual void reconcile(vdom const &) { }
+};
+
+class vdom {
+private:
+  vdom_cell * m_ptr;
+  friend class vdom_cell;
+public:
+  vdom(vdom_cell * ptr) : m_ptr(ptr) { lean_assert(m_ptr);  m_ptr->inc_ref(); }
+  vdom(vdom const & s) : m_ptr(s.m_ptr) { if (m_ptr) m_ptr->inc_ref(); }
+  vdom(vdom && s) : m_ptr(s.m_ptr) { s.m_ptr = nullptr; }
+  ~vdom() {if (m_ptr) m_ptr->dec_ref(); }
+  vdom & operator=(vdom const & s) { LEAN_COPY_REF(s); }
+  vdom & operator=(vdom && s) { LEAN_MOVE_REF(s); }
+  void reconcile(vdom const & old_vdom) { m_ptr->reconcile(old_vdom); }
+  vdom_cell * raw() const { return m_ptr; }
+  json to_json(list<unsigned> const & route = {}) const { return m_ptr->to_json(route); }
+  optional<std::string> key() { return m_ptr->key(); }
+};
+
+struct vdom_element : public vdom_cell {
+    std::string m_tag;
+    json m_attrs;
+    std::map<std::string, unsigned> m_events;
+    std::vector<vdom> m_children;
+    optional<vdom> m_tooltip;
+    vdom_element(std::string const & tag, json const & attrs, std::map<std::string, unsigned> events, std::vector<vdom> const & children, optional<vdom> const & tooltip)
+      : m_tag(tag), m_attrs(attrs), m_events(events), m_children(children), m_tooltip(tooltip) {}
+    optional<std::string> key() override;
+    void reconcile(vdom const & old_vdom) override;
+    json to_json(list<unsigned> const & route) override;
+};
+
+struct vdom_string : public vdom_cell {
+    std::string m_val;
+    vdom_string(std::string const & val) : m_val(val) {}
+    json to_json(list<unsigned> const &) override { return m_val; }
+};
+
+class component_instance : public vdom_cell {
+    ts_vm_obj const m_component;
+    unsigned int m_component_hash;
+
+    ts_vm_obj m_props;
+    optional<ts_vm_obj> m_state;
+
+    std::vector<vdom> m_render;
+    // [note] these component instances are owned by one of the vdoms in `m_render`.
+    // Should never point to something not owned by m_render.
+    std::vector<component_instance *> m_children;
+    std::map<unsigned, ts_vm_obj> m_handlers;
+
+    unsigned m_id;
+    list<unsigned> m_route;
+
+    bool m_has_rendered;
+    unsigned m_reconcile_count;
+
+    vm_obj init(vm_obj const & p, optional<vm_obj> const & s);
+    pair<vm_obj, optional<vm_obj>> update(vm_obj const & p, vm_obj const & s, vm_obj const & a);
+    vm_obj view(vm_obj const & p, vm_obj const & s);
+    bool props_are_equal(vm_obj const & p_old, vm_obj const & p_new);
+public:
+    void render();
+    component_instance(vm_obj const & c, vm_obj const & props, list<unsigned> const & route = list<unsigned>()) : m_component(c), m_props(props), m_route(route) {
+      static unsigned count = 0; // [fixme] need to worry about thread safety here. use a global pointer or something.
```

Comment:
Definitely a fixme.

Reply:
I don't know what to do here...

Reply:
I would add a global variable `std::atomic_uint g_fresh_component_instance_id` in `widget.cpp` and then call `g_fresh_component_instance_id.fetch_add(1)`.  https://en.cppreference.com/w/cpp/atomic/atomic/fetch_add